File: 2023/Day 2/perl/day2.pl
#!/usr/bin/perl
use strict;
use warnings;

sub day2 {
    my ($data, $cube_limits) = @_;
    my $sum_of_possible_game_ids = 0;
    my $total_power = 0;

    foreach my $game (@$data) {
        my ($game_id_str, $rounds_info) = split /: /, $game;
        my @game_id_parts = split / /, $game_id_str;
        my $game_id = pop @game_id_parts;

        my @rounds = split /; /, $rounds_info;
        my %max_cubes = (red => 0, green => 0, blue => 0);
        my $possible = 1;

        foreach my $round (@rounds) {
            my @scores = split /, /, $round;
            my %round_counts = (red => 0, green => 0, blue => 0);

            foreach my $score (@scores) {
                my ($value, $colour) = split / /, $score;
                $round_counts{$colour} = $value;
                $max_cubes{$colour} = $max_cubes{$colour} > $round_counts{$colour} ? $max_cubes{$colour} : $round_counts{$colour};
            }
        }

        foreach my $colour (keys %max_cubes) {
            if ($max_cubes{$colour} > $cube_limits->{$colour}) {
                $possible = 0;
                last;
            }
        }

        $sum_of_possible_game_ids += $game_id if $possible;

        $total_power += $max_cubes{red} * $max_cubes{green} * $max_cubes{blue};
    }

    return ($sum_of_possible_game_ids, $total_power);
}


sub main {
    my $filename = "input.txt";
    open my $fh, '<', $filename or die "Could not open file: $!";
    chomp(my @data = <$fh>);
    close $fh;
    my $cube_limits = { red => 12, green => 13, blue => 14 };

    my ($sum_of_ids, $sum_of_powers) = day2(\@data, $cube_limits);
    print "Sum of IDs: $sum_of_ids, Total Power: $sum_of_powers\n";
}

main();

File: 2023/Day 2/php/day2.php
<?php

function day2($data, $cube_limits) {
    $sum_of_possible_game_ids = 0;
    $total_power = 0;

    foreach ($data as $game) {
        list($game_id_str, $rounds_info) = explode(": ", $game);
        $game_id_parts = explode(" ", $game_id_str);
        $game_id = intval(end($game_id_parts));

        $rounds = explode("; ", $rounds_info);
        $max_cubes = array("red" => 0, "green" => 0, "blue" => 0);
        $possible = true;

        foreach ($rounds as $round) {
            $scores = explode(", ", $round);
            $round_counts = array("red" => 0, "green" => 0, "blue" => 0);

            foreach ($scores as $score) {
                list($value, $colour) = explode(" ", $score);
                $round_counts[$colour] = intval($value);
                $max_cubes[$colour] = max($max_cubes[$colour], $round_counts[$colour]);
            }
        }

        // Check if the game is possible
        foreach ($max_cubes as $colour => $count) {
            if ($count > $cube_limits[$colour]) {
                $possible = false;
                break;
            }
        }

        // Part 1
        if ($possible) {
            $sum_of_possible_game_ids += $game_id;
        }

        // Part 2: Calculate total power regardless of whether the game is possible
        $total_power += $max_cubes["red"] * $max_cubes["green"] * $max_cubes["blue"];
    }

    return array($sum_of_possible_game_ids, $total_power);
}

$filename = "input.txt";
$data = file($filename, FILE_IGNORE_NEW_LINES | FILE_SKIP_EMPTY_LINES);
$cube_limits = array("red" => 12, "green" => 13, "blue" => 14);

list($sum_of_ids, $sum_of_powers) = day2($data, $cube_limits);
echo "Sum of IDs: $sum_of_ids, Total Power: $sum_of_powers";
?>